rainier-core/src/main/scala/com/stripe/rainier/compute/Context.scala

```diff
@@ -1,10 +1,71 @@
 package com.stripe.rainier.compute
 
-case class Context(density: Real) {
+case class Context(base: Real, batched: Target) {
   val compiler: Compiler = IRCompiler(200, 100)
-  val variables: List[Variable] = RealOps.variables(density).toList
-  lazy val gradient: List[Real] = Gradient.derive(variables, density).toList
 
-  def compileDensity: Array[Double] => Double =
-    compiler.compile(variables, density)
+  val variables: List[Variable] = {
+    val allVariables =
+      RealOps.variables(base) ++
+        RealOps.variables(batched.real) --
+        batched.placeholders.keySet
```

Comment:
The original `real` in the `Target`, and hence the `newReal` in the batched `Target`, can also refer to `variables` that are not placeholders (ie, that are true parameters). These are the ones we want (because they're the ones we care about the gradients for).

Reply:
aaah, got it.